### ISSUE TITLE
QEMU backend fails to realize dnsmasq died (and doesn't restart it)

### ISSUE THREAD

**Issue by markshuttle** [canonical/multipass#1475]
On my Ubuntu 20.04 laptop I am unable to launch new instances:

mark@thrifty:~$ multipass launch ubuntu
launch failed: The following errors occurred:                                   
memorable-topminnow: timed out waiting for response
mark@thrifty:~$ multipass launch ubuntu                                         
launch failed: The following errors occurred:                                   
arriving-rabbit: timed out waiting for response

Interestingly, qemu is still running for both attempts:

mark@thrifty:~$ ps ax | grep qemu                                               
1904243 ?        Sl     0:21 /snap/multipass/1784/usr/bin/qemu-system-x86_64 --enable-kvm -device virtio-scsi-pci,id=scsi0 -drive file=/var/snap/multipass/common/data/multipassd/vault/instances/memorable-topminnow/ubuntu-18.04-server-cloudimg-amd64.img,if=none,format=qcow2,discard=unmap,id=hda -device scsi-hd,drive=hda,bus=scsi0.0 -smp 1 -m 1024M -device virtio-net-pci,netdev=hostnet0,id=net0,mac=52:54:00:e0:74:c4 -netdev tap,id=hostnet0,ifname=tap-97e3bbd7047,script=no,downscript=no -qmp stdio -cpu host -chardev null,id=char0 -serial chardev:char0 -nographic -cdrom /var/snap/multipass/common/data/multipassd/vault/instances/memorable-topminnow/cloud-init-config.iso
1904495 ?        Sl     0:20 /snap/multipass/1784/usr/bin/qemu-system-x86_64 --enable-kvm -device virtio-scsi-pci,id=scsi0 -drive file=/var/snap/multipass/common/data/multipassd/vault/instances/arriving-rabbit/ubuntu-18.04-server-cloudimg-amd64.img,if=none,format=qcow2,discard=unmap,id=hda -device scsi-hd,drive=hda,bus=scsi0.0 -smp 1 -m 1024M -device virtio-net-pci,netdev=hostnet0,id=net0,mac=52:54:00:af:05:38 -netdev tap,id=hostnet0,ifname=tap-c854facc299,script=no,downscript=no -qmp stdio -cpu host -chardev null,id=char0 -serial chardev:char0 -nographic -cdrom /var/snap/multipass/common/data/multipassd/vault/instances/arriving-rabbit/cloud-init-config.iso

**Logs**
Apr 07 20:05:49 thrifty multipassd[1029]: Updating 20.04 source image to latest
Apr 07 20:05:49 thrifty multipassd[1029]:   0%
Apr 07 20:06:02 thrifty multipassd[1029]:   10%
Apr 07 20:06:13 thrifty multipassd[1029]:   20%
Apr 07 20:06:32 thrifty multipassd[1029]:   30%
Apr 07 20:06:46 thrifty multipassd[1029]:   40%
Apr 07 20:06:57 thrifty multipassd[1029]:   50%
Apr 07 20:07:08 thrifty multipassd[1029]:   60%
Apr 07 20:07:18 thrifty multipassd[1029]:   70%
Apr 07 20:07:29 thrifty multipassd[1029]:   80%
Apr 07 20:07:39 thrifty multipassd[1029]:   90%
Apr 07 20:07:51 thrifty multipassd[1029]:   100%
Apr 07 20:07:55 thrifty multipassd[1874915]: ESC[0;38;5;245mESC[0;1;31mESC[0;38;5;245mApplying AppArmor policy: multipass.qemu-imgESC[0m
Apr 08 02:05:48 thrifty multipassd[1029]: ESC[0;38;5;245mESC[0;1;31mESC[0;38;5;245mChecking for images to update…ESC[0m
Apr 08 08:05:48 thrifty multipassd[1029]: ESC[0;38;5;245mESC[0;1;31mESC[0;38;5;245mChecking for images to update…ESC[0m
Apr 08 08:34:21 thrifty multipassd[1904175]: ESC[0;38;5;245mESC[0;1;31mESC[0;38;5;245mApplying AppArmor policy: multipass.qemu-imgESC[0m
Apr 08 08:34:22 thrifty multipassd[1904184]: ESC[0;38;5;245mESC[0;1;31mESC[0;38;5;245mApplying AppArmor policy: multipass.qemu-imgESC[0m
Apr 08 08:34:22 thrifty multipassd[1904190]: ESC[0;38;5;245mESC[0;1;31mESC[0;38;5;245mApplying AppArmor policy: multipass.qemu-imgESC[0m
Apr 08 08:34:22 thrifty multipassd[1029]: ESC[0;38;5;245mESC[0;1;31mESC[0;38;5;245mprocess working dir '/snap/multipass/1784/qemu'ESC[0m
Apr 08 08:34:22 thrifty multipassd[1029]: process program 'qemu-system-x86_64'
Apr 08 08:34:22 thrifty multipassd[1029]: process arguments '--enable-kvm, -device, virtio-scsi-pci,id=scsi0, -drive, file=/var/snap/multipass/common/data/multipassd/vault/instances/memorable-topminnow/ubuntu-18.04-server-cloudimg-amd64.img,if=none,format=qcow2,discard=unmap,id=hda, -device, scsi-hd,drive=hda,bus=scsi0.0, -smp, 1, -m, 1024M, -device, virtio-net-pci,netdev=hostnet0,id=net0,mac=52:54:00:e0:74:c4, -netdev, tap,id=hostnet0,ifname=tap-97e3bbd7047,script=no,downscript=no, -qmp, stdio, -cpu, host, -chardev, null,id=char0, -serial, chardev:char0, -nographic, -cdrom, /var/snap/multipass/common/data/multipassd/vault/instances/memorable-topminnow/cloud-init-config.iso'
Apr 08 08:34:22 thrifty multipassd[1029]: ESC[0;1;38;5;185mESC[0;1;39mESC[0;1;38;5;185mqemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.vmx [bit 5]ESC[0m
Apr 08 08:34:23 thrifty multipassd[1904243]: ESC[0;38;5;245mESC[0;1;31mESC[0;38;5;245mApplying AppArmor policy: multipass.memorable-topminnow.qemu-system-x86_64ESC[0m
Apr 08 08:34:23 thrifty multipassd[1029]: process started
Apr 08 08:34:23 thrifty multipassd[1029]: ESC[0;38;5;245mESC[0;1;31mESC[0;38;5;245mQMP: {"QMP": {"version": {"qemu": {"micro": 1, "minor": 11, "major": 2}, "package": "(Debian 1:2.11+dfsg-1ubuntu7.23)"}, "capabilities": []}}ESC[0m
Apr 08 08:34:23 thrifty multipassd[1029]: ESC[0;38;5;245mESC[0;1;31mESC[0;38;5;245mQMP: {"return": {}}ESC[0m
Apr 08 08:34:40 thrifty multipassd[1029]: ESC[0;38;5;245mESC[0;1;31mESC[0;38;5;245mQMP: {"timestamp": {"seconds": 1586331280, "microseconds": 459818}, "event": "NIC_RX_FILTER_CHANGED", "data": {"name": "net0", "path": "/machine/peripheral/net0/virtio-backend"}}ESC[0m
Apr 08 08:36:32 thrifty multipassd[1904431]: ESC[0;38;5;245mESC[0;1;31mESC[0;38;5;245mApplying AppArmor policy: multipass.qemu-imgESC[0m
Apr 08 08:36:32 thrifty multipassd[1904437]: ESC[0;38;5;245mESC[0;1;31mESC[0;38;5;245mApplying AppArmor policy: multipass.qemu-imgESC[0m
Apr 08 08:36:33 thrifty multipassd[1029]: ESC[0;38;5;245mESC[0;1;31mESC[0;38;5;245mprocess working dir '/snap/multipass/1784/qemu'ESC[0m
Apr 08 08:36:33 thrifty multipassd[1029]: process program 'qemu-system-x86_64'
Apr 08 08:36:33 thrifty multipassd[1029]: process arguments '--enable-kvm, -device, virtio-scsi-pci,id=scsi0, -drive, file=/var/snap/multipass/common/data/multipassd/vault/instances/arriving-rabbit/ubuntu-18.04-server-cloudimg-amd64.img,if=none,format=qcow2,discard=unmap,id=hda, -device, scsi-hd,drive=hda,bus=scsi0.0, -smp, 1, -m, 1024M, -device, virtio-net-pci,netdev=hostnet0,id=net0,mac=52:54:00:af:05:38, -netdev, tap,id=hostnet0,ifname=tap-c854facc299,script=no,downscript=no, -qmp, stdio, -cpu, host, -chardev, null,id=char0, -serial, chardev:char0, -nographic, -cdrom, /var/snap/multipass/common/data/multipassd/vault/instances/arriving-rabbit/cloud-init-config.iso'
Apr 08 08:36:33 thrifty multipassd[1029]: ESC[0;1;38;5;185mESC[0;1;39mESC[0;1;38;5;185mqemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.vmx [bit 5]ESC[0m
Apr 08 08:36:33 thrifty multipassd[1904495]: ESC[0;38;5;245mESC[0;1;31mESC[0;38;5;245mApplying AppArmor policy: multipass.arriving-rabbit.qemu-system-x86_64ESC[0m
Apr 08 08:36:33 thrifty multipassd[1029]: process started
Apr 08 08:36:33 thrifty multipassd[1029]: ESC[0;38;5;245mESC[0;1;31mESC[0;38;5;245mQMP: {"QMP": {"version": {"qemu": {"micro": 1, "minor": 11, "major": 2}, "package": "(Debian 1:2.11+dfsg-1ubuntu7.23)"}, "capabilities": []}}ESC[0m
Apr 08 08:36:33 thrifty multipassd[1029]: ESC[0;38;5;245mESC[0;1;31mESC[0;38;5;245mQMP: {"return": {}}ESC[0m
Apr 08 08:36:49 thrifty multipassd[1029]: ESC[0;38;5;245mESC[0;1;31mESC[0;38;5;245mQMP: {"timestamp": {"seconds": 1586331409, "microseconds": 878531}, "event": "NIC_RX_FILTER_CHANGED", "data": {"name": "net0", "path": "/machine/peripheral/net0/virtio-backend"}}ESC[0m

mark@thrifty:~$ multipass info --all
Name:           arriving-rabbit
State:          Unknown
IPv4:           --
Release:        --
Image hash:     2f6bc5e7d9ac (Ubuntu 18.04 LTS)
Load:           --
Disk usage:     --
Memory usage:   --

Name:           memorable-topminnow
State:          Unknown
IPv4:           --
Release:        --
Image hash:     2f6bc5e7d9ac (Ubuntu 18.04 LTS)
Load:           --
Disk usage:     --
Memory usage:   --

**Comment by Saviq** [comment]
Hi @markshuttle the instances never got IP addresses. Can you please check if `dnsmasq` is running under "CGroup:" in `systemctl status snap.multipass.multipassd.service`:

```console
$ systemctl status snap.multipass.multipassd.service | grep -A2 CGroup
     CGroup: /system.slice/snap.multipass.multipassd.service
             ├─199163 /snap/multipass/1784/bin/multipassd --verbosity debug --logger platform
             └─199399 /snap/multipass/1784/usr/sbin/dnsmasq --strict-order --bind-interfaces --pid-file=/var/snap/multipass/common/data/multipassd/network/dnsmasq.pid --domain=multipass --local=/multipass/ --except-interface=lo --interface=mpqemubr0 --listen-address=10.65.118.1 --dhcp-no-override --dhcp-authoritative --dhcp-leasefile=/var/snap/multipass/common/data/multipassd/network/dnsmasq.leases --dhcp-hostsfile=/var/snap/multipass/common/data/multipassd/network/dnsmasq.hosts --dhcp-range 10.65.118.2,10.65.118.254,infinite
```

And, if not, check its logs for why it failed:
```console
$ journalctl -u snap.multipass.multipassd.service _COMM=dnsmasq
```

If you have a `/etc/dnsmasq.conf`, we just fixed #1468, so please try refreshing to edge.

If none of that suggests anything, do you have a firewall running that could be blocking the DHCP requests between instances and dnsmasq?

**Comment by markshuttle** [comment]
Hiya. I don't have /etc/dnsmasq.conf so it's not #1468 

Looks like dnsmasq is not running:

$ systemctl status snap.multipass.multipassd.service | grep -A2 CGroup
     CGroup: /system.slice/snap.multipass.multipassd.service
             └─1029 /snap/multipass/1784/bin/multipassd --verbosity debug --logger platform

As for logs, I don't have anything for April which is interesting. My last logs are quite old:

Mar 18 10:18:36 thrifty dnsmasq[1216]: started, version 2.79 cachesize 150
Mar 18 10:18:36 thrifty dnsmasq[1216]: compile time options: IPv6 GNU-getopt DBus i18n IDN DHCP DHCPv6 no-Lua TFTP conntrack ipset auth DNSSEC loop-detect inotify
Mar 18 10:18:36 thrifty dnsmasq-dhcp[1216]: DHCP, IP range 10.50.117.2 -- 10.50.117.254, lease time infinite
Mar 18 10:18:36 thrifty dnsmasq-dhcp[1216]: DHCP, sockets bound exclusively to interface mpqemubr0
Mar 18 10:18:36 thrifty dnsmasq[1216]: using local addresses only for domain multipass
Mar 18 10:18:36 thrifty dnsmasq[1216]: reading /etc/resolv.conf
Mar 18 10:18:36 thrifty dnsmasq[1216]: using local addresses only for domain multipass
Mar 18 10:18:36 thrifty dnsmasq[1216]: using nameserver 127.0.0.53#53
Mar 18 10:18:36 thrifty dnsmasq[1216]: read /etc/hosts - 25 addresses
Mar 18 10:18:36 thrifty dnsmasq[1216]: cannot read /var/snap/multipass/common/data/multipassd/network/dnsmasq.hosts: No such file or directory
Mar 18 10:18:41 thrifty dnsmasq[1216]: reading /etc/resolv.conf
Mar 18 10:18:41 thrifty dnsmasq[1216]: using local addresses only for domain multipass
Mar 18 10:18:41 thrifty dnsmasq[1216]: using nameserver 127.0.0.53#53
Mar 18 10:18:41 thrifty dnsmasq[1216]: reading /etc/resolv.conf
Mar 18 10:18:41 thrifty dnsmasq[1216]: using local addresses only for domain multipass
Mar 18 10:18:41 thrifty dnsmasq[1216]: using nameserver 127.0.0.53#53
Mar 18 10:18:41 thrifty dnsmasq[1216]: reading /etc/resolv.conf
Mar 18 10:18:41 thrifty dnsmasq[1216]: using local addresses only for domain multipass
Mar 18 10:18:41 thrifty dnsmasq[1216]: using nameserver 127.0.0.53#53
Mar 18 10:18:41 thrifty dnsmasq[1216]: reading /etc/resolv.conf
Mar 18 10:18:41 thrifty dnsmasq[1216]: using local addresses only for domain multipass
Mar 18 10:18:41 thrifty dnsmasq[1216]: using nameserver 127.0.0.53#53
Mar 18 10:18:41 thrifty dnsmasq[1216]: reading /etc/resolv.conf
Mar 18 10:18:41 thrifty dnsmasq[1216]: using local addresses only for domain multipass
Mar 18 10:18:41 thrifty dnsmasq[1216]: using nameserver 127.0.0.53#53
Mar 18 10:18:41 thrifty dnsmasq[1216]: reading /etc/resolv.conf
Mar 18 10:18:41 thrifty dnsmasq[1216]: using local addresses only for domain multipass
Mar 18 10:18:41 thrifty dnsmasq[1216]: using nameserver 127.0.0.53#53

**Comment by markshuttle** [comment]
Restarting the multipass snap started dnsmasq.Launching instances works now, but I think we need to harden multipass against dnsmasq failures. For example:

 - the launch command should have failed fast saying that dnsmasq wasn't running
 - we should have some sort of auto-restart piece

**Comment by Saviq** [comment]
> As for logs, I don't have anything for April which is interesting. My last logs are quite old:

That's interesting indeed, how about `journalctl -u snap.multipass.multipassd.service` without the `_COMM`?

> * the launch command should have failed fast saying that dnsmasq wasn't running
> * we should have some sort of auto-restart piece

Yes indeed, we need to add a health check for that.

**Comment by townsend2010** [comment]
We do have a health check in place, but a flaw in how the check for determining if `dnsmasq` is running was exposed here.  After some discussion on IRC, I think there is a better way we can handle this and will work on fixing that.

Thanks!

**Comment by townsend2010** [comment]
This issue should have been solved in #1748.  We can reopen or open a new issue if this is seen again.

**Comment by daniellenzofficial** [comment]
> This issue should have been solved in #1748. We can reopen or open a new issue if this is seen again.

Wouldn't it be nice idea to the community to use the edge version?

**Comment by townsend2010** [comment]
This should have been closed a long time ago since it's been fixed for a while:)
